src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -21237,42 +21302,76 @@ public class CStruct<TStruct> where TStruct : struct
             var lib = CreateNullableCompilation(new[] { MaybeNullAttributeDefinition, lib_cs });
 
             var source = @"
+using System.Diagnostics.CodeAnalysis;
 class C
 {
-    static void M1<T>(T t1)
+    static void M1<T>([MaybeNull]T t1)
     {
-        new COpen<T>().f1.ToString(); // 1
+        new COpen<T>().f1 = t1;
+        new COpen<T>().f1.ToString(); // 0, 1
+
+        var xOpen = new COpen<T>();
+        xOpen.f1 = t1;
+        xOpen.f1.ToString(); // 2, 3
+
+        var xOpen2 = new COpen<T>();
+        xOpen2.f1.ToString(); // 4, 5
+        xOpen2.f1.ToString();
+
+        var xOpen3 = new COpen<T>();
+        if (xOpen3.f1 is object) // 6
+        {
+            xOpen3.f1.ToString();
```

Comment:
I'm slightly surprised we knew we didn't need to emit a warning here, but I'm not complaining

Reply:
Not producing a warning here is very useful indeed, otherwise we'd produce the same warning over and over again even after you checked for null.
This is achieved by not reporting the warning for [MaybeNull]T after it was assigned a definitely not-null state (we check the state in `ReportMaybeNullFromTypeParameterValueIfNeeded`)

---
In reply to: [324902033](https://github.com/dotnet/roslyn/pull/38610#discussion_r324902033) [](ancestors = 324902033)